packages/babel-generator/src/buffer.ts

```diff
@@ -87,50 +151,105 @@ export default class Buffer {
+    this._flush();
+    this._appendChar(char, 1, this._sourcePosition);
   }
 
   /**
    * Add a string to the buffer than can be reverted.
    */
-  queue(str: string): void {
+  queue(char: number): void {
     // Drop trailing spaces when a newline is inserted.
-    if (str === "\n") {
-      while (this._queue.length > 0 && SPACES_RE.test(this._queue[0][0])) {
-        this._queue.shift();
+    if (char === charcodes.lineFeed) {
+      while (this._queueCursor !== 0) {
+        const char = this._queue[this._queueCursor - 1].char;
+        if (char !== charcodes.space && char !== charcodes.tab) {
+          break;
+        }
+
+        this._queueCursor--;
       }
     }
 
-    const { line, column, filename, identifierName } = this._sourcePosition;
-    this._queue.unshift([str, line, column, identifierName, filename]);
+    const sourcePosition = this._sourcePosition;
+    this._pushQueue(
+      char,
+      1,
+      sourcePosition.line,
+      sourcePosition.column,
+      sourcePosition.identifierName,
+      sourcePosition.filename,
+    );
   }
 
   /**
    * Same as queue, but this indentation will never have a sourcmap marker.
    */
-  queueIndentation(str: string): void {
-    this._queue.unshift([str, undefined, undefined, undefined, undefined]);
+  queueIndentation(char: number, repeat: number): void {
+    this._pushQueue(char, repeat, undefined, undefined, undefined, undefined);
   }
 
   _flush(): void {
-    let item: Parameters<Buffer["_append"]>;
-    while ((item = this._queue.pop())) {
-      this._append(...item);
+    const queueCursor = this._queueCursor;
+    const queue = this._queue;
+    for (let i = 0; i < queueCursor; i++) {
+      const item: queueObj = queue[i];
+      this._appendChar(item.char, item.repeat, item);
     }
+    this._queueCursor = 0;
   }
 
-  _append(
-    str: string,
-    line: number | undefined,
-    column: number | undefined,
-    identifierName: string | undefined,
-    filename: string | undefined,
-  ): void {
-    this._buf += str;
-    this._last = str.charCodeAt(str.length - 1);
+  _appendChar(char: number, repeat: number, sourcePos: SourcePos): void {
+    this._last = char;
+
+    this._str +=
+      repeat > 1
+        ? String.fromCharCode(char).repeat(repeat)
+        : String.fromCharCode(char);
+
+    if (char !== charcodes.lineFeed) {
+      this._mark(
+        sourcePos.line,
+        sourcePos.column,
+        sourcePos.identifierName,
+        sourcePos.filename,
+      );
+      this._position.column += repeat;
+    } else {
+      this._position.line++;
+      this._position.column = 0;
+    }
+  }
+
+  _append(str: string, sourcePos: SourcePos, maybeNewline: boolean): void {
+    const len = str.length;
+
+    this._last = str.charCodeAt(len - 1);
+
+    if (++this._appendCount > 4096) {
+      // @ts-ignore
+      this._str | 0; // Unexplainable huge performance boost. Ref: https://github.com/davidmarkclements/flatstr License: MIT
```

Comment:
The repo offered an explanation about how this approach works for V8: https://github.com/davidmarkclements/flatstr#how-does-it-work

It triggers V8's flat string via side effects. Because it is an engine specific hack, in the source, the author suggests relying on the package instead of copy-pasting it:

```
// You may be tempted to copy and paste this, 
// but take a look at the commit history first,
// this is a moving target so relying on the module
// is the best way to make sure the optimization
// method is kept up to date` and compatible with
// every Node version.
```

Though I think `+this._str` should do the trick, with 3 bytes shorter than ` | 0`.

For reference, here are my local benchmark results:

without this hack:
```
node --predictable ./real-case/jquery.mjs
baseline 1 jquery 3.6: 64 ops/sec ±10.45% (16ms)
baseline 4 jquery 3.6: 13.68 ops/sec ±1.06% (73ms)
baseline 16 jquery 3.6: 2.81 ops/sec ±26.7% (355ms)
baseline 64 jquery 3.6: 0.75 ops/sec ±6.47% (1330ms)
current 1 jquery 3.6: 95.27 ops/sec ±10.87% (10ms)
current 4 jquery 3.6: 14.31 ops/sec ±11.89% (70ms)
current 16 jquery 3.6: 3.26 ops/sec ±22.33% (307ms)
current 64 jquery 3.6: 0.89 ops/sec ±16.89% (1119ms)
```

`+` approach:
```
$ node --predictable ./real-case/jquery.mjs
baseline 1 jquery 3.6: 67.18 ops/sec ±10.51% (15ms)
baseline 4 jquery 3.6: 14.05 ops/sec ±0.84% (71ms)
baseline 16 jquery 3.6: 2.77 ops/sec ±31.2% (361ms)
baseline 64 jquery 3.6: 0.79 ops/sec ±3.29% (1268ms)
current 1 jquery 3.6: 92.05 ops/sec ±15.16% (11ms)
current 4 jquery 3.6: 22.96 ops/sec ±15.59% (44ms)
current 16 jquery 3.6: 5.91 ops/sec ±19.89% (169ms)
current 64 jquery 3.6: 1.59 ops/sec ±1.73% (630ms)
```

` | 0` approach:
```
baseline 1 jquery 3.6: 65.05 ops/sec ±10.47% (15ms)
baseline 4 jquery 3.6: 13.58 ops/sec ±1.07% (74ms)
baseline 16 jquery 3.6: 3.08 ops/sec ±7.63% (325ms)
baseline 64 jquery 3.6: 0.72 ops/sec ±20.03% (1391ms)
current 1 jquery 3.6: 93.78 ops/sec ±10.35% (11ms)
current 4 jquery 3.6: 21.35 ops/sec ±15.54% (47ms)
current 16 jquery 3.6: 6.28 ops/sec ±0.85% (159ms)
current 64 jquery 3.6: 1.4 ops/sec ±28.66% (716ms)
```

Notice how the `flatstr` improves the performance when generating large files ( > = 4 jquery ).

Reply:
Yes, I know this should have something to do with Flatten, but it's still weird.
At first I used `Number(str)`, and later referenced `str | 0` in this package, so I didn't refer to this package directly.
The benchmark results are amazing, we can avoid this hack with a simple `if` in the buffer small enough!
In addition, your benchmark seems to have a large floating range. Perhaps you should avoid running other programs that occupy the CPU at the same time. At least on my computer, the benchmark is quite unstable, and even produces a 10% error, depending on which physical core is scheduled, which is really helpless.

PS: Your computer is really fast. :)

Reply:
I've only made minor changes at the moment, the small files I'll do later when I have time.

Reply:
Ah, I seem to be reading something wrong, it seems that this hack has no effect on small buffers.

Reply:
Ah, it was because I didn't update the dependencies. Should I change `@babel/preset-typescript` to `workspace:^`?